### PR TITLE
Add orchestrator version support in vlabs API model and enable creation of versioned Kubernetes clusters

### PR DESCRIPTION
**Pull request by amanohar**
- Introduce orchestrator version support in vlabs API model and enable creation of versioned Kubernetes clusters
- Add orchestrator version tag to master and agent VMs
- Pin Swarm Mode agent versions
- Fix windowsProfile validation bug

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/541)

### CODE REVIEW

**Comment by amanohar** [comment]
@acs-bot test this please

File: pkg/api/vlabs/validate.go [hunk 1]
```diff
@@ -13,10 +13,25 @@ func (o *OrchestratorProfile) Validate() error {
...
 	case DCOS184:
+		if e := o.ValidateOrchestratorVersion(DCOS184Version); e != nil {
+			return e
+		}
 	case DCOS173:
+		if e := o.ValidateOrchestratorVersion(DCOS173Version); e != nil {
+			return e
+		}
 	case Swarm:
 	case Kubernetes:
```
Comment (rjtsdl) [inline pkg/api/vlabs/validate.go]: Do we need to add Kubernetes validateOrchestratorVersion?

Comment (amanohar) [inline pkg/api/vlabs/validate.go]: Its not needed. This check is to make sure that DCOS orchestrator and version is aligned.

**Comment by amanohar** [comment]
@acs-bot test this please

File: pkg/api/const.go [hunk 2]
```diff
@@ -59,3 +59,33 @@ const (
 	// ManagedDisks means that the nodes use managed disks for their os and attached volumes
 	ManagedDisks = "ManagedDisks"
 )
+
+// KubernetesVersion defines supported Kubernetes versions
+type KubernetesVersion string
```
Comment (JiangtianLi) [inline pkg/api/const.go]: Is KubernetesVersion used anywhere? Seems OrchestratorVersion is sufficient?

File: pkg/api/const.go [hunk 3]
```diff
@@ -59,3 +59,33 @@ const (
...
+
+// DCOSVersion defines supported Kubernetes versions
+type DCOSVersion string
```
Comment (JiangtianLi) [inline pkg/api/const.go]: The same as above. Seems it is obsolete?

File: pkg/api/vlabs/validate.go [hunk 4]
```diff
@@ -32,6 +47,15 @@ func (o *OrchestratorProfile) Validate() error {
 	return nil
 }
 
+// ValidateOrchestratorVersion validates orchestrator version
+func (o *OrchestratorProfile) ValidateOrchestratorVersion(orchestratorVersion OrchestratorVersion) error {
```
Comment (JiangtianLi) [inline pkg/api/vlabs/validate.go]: Do we validate version against existing versions? E.g. if orchestratortype is Kubernetes, then only 1.6.2, 1.6.0, and 1.5.3 is supported?

File: pkg/api/convertertoapi.go [hunk 5]
```diff
@@ -376,13 +379,57 @@ func convertV20160330OrchestratorProfile(v20160330 *v20160330.OrchestratorProfil
 
 func convertV20170131OrchestratorProfile(v20170131 *v20170131.OrchestratorProfile, api *OrchestratorProfile) {
 	api.OrchestratorType = OrchestratorType(v20170131.OrchestratorType)
+	if api.OrchestratorType == Kubernetes {
+		api.OrchestratorVersion = Kubernetes162
```
Comment (JiangtianLi) [inline pkg/api/convertertoapi.go]: Right now, the testdata json and example json all take default OrchestratorVersion? Can we update some so the user has the example of input?

File: pkg/api/vlabs/validate.go (hunk 4 above)
Comment (amanohar) [inline pkg/api/vlabs/validate.go]: Added.

File: pkg/api/convertertoapi.go (hunk 5 above)
Comment (amanohar) [inline pkg/api/convertertoapi.go]: I was waiting to write tests for nightly Jenkins run. Added now.

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by amanohar** [comment]
@acs-bot test this please

File: pkg/api/vlabs/types.go [hunk 6]
```diff
@@ -202,6 +203,9 @@ type KeyVaultCertificate struct {
 // OrchestratorType defines orchestrators supported by ACS
 type OrchestratorType string
 
+// OrchestratorVersion defines the version for orchestratorType
+type OrchestratorVersion string
```
Comment (anhowe) [inline pkg/api/vlabs/types.go]: to detect serialization / deserialization issues, please add some version tests with a variety of versions to pkg/acsengine/testdata/

**Comment by dmitsh** [comment]
@acs-bot test this please

**Comment by dmitsh** [comment]
@acs-bot test this please

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by dmitsh** [comment]
@acs-bot test this please

**Comment by JackQuincy** [comment]
Seems like the DCOS changes haven't been finished. You can set an orchestrator version but it doesn't make anything change or happen in the engine/templates.

---

Review status: 0 of 57 files reviewed at latest revision, 7 unresolved discussions.

---

*[parts/kubernetesmasterresources.t, line 361 at r4](https://reviewable.io:443/reviews/azure/acs-engine/541#-Kiq45vyvYncsz86uVOI:-Kiq45vyvYncsz86uVOJ:bpml0su) ([raw file](https://github.com/azure/acs-engine/blob/7b4f7e84c0e384c9d4bf62a8ad9f4abb84eb0fa1/parts/kubernetesmasterresources.t#L361)):*
> {{.OrchestratorProfile.OrchestratorType}}:{{.OrchestratorProfile.OrchestratorVersion}}

why not use the variable you created
`[variables('orchestratorNameVersionTag')]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/acs-engine/541#-:-Kiq6bGjML0hGYOkPbzE:bea1600)*

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by amanohar** [comment]
@acs-bot test this please

**Pull request merged**